### ISSUE TITLE
quantify_output problems with sample names (?) + cannot count empty files as human reads

### ISSUE THREAD

**Issue by thierryjanssens** [DennisSchmitz/Jovian_archive#57]
Jovian encounters difficulties in a part of the quantify_output rule. Bad read sums are encountered.
(The data were generated on eight samples by means of SISPA and subsequent Nextseq 2x150 bp sequencing). The error went as follows:

```

Now counting the number of lines in the fastq files from which
reads have been discarded that were mapped to the human genome 
by Bowtie2 to determine the number of non-human reads.

This may take a while...

Done counting!===============================================] 100.0% - Reading files  [ 24 / 24 ]
Now checking if the numbers add up properly...
Error! Sample 4311801120_S23_R1_ has a bad reads sum: 3434774 > total (3384998)
Error! Sample 4311801213_S7_R1_ has a bad reads sum: 11845414 > total (11003928)
Error! Sample 4311801221_S16_R1_ has a bad reads sum: 4909565 > total (4898292)
Error! Sample 4311801325_S19_R1_ has a bad reads sum: 6974119 > total (4968376)
Error! Sample 4311900015_S21_R1_ has a bad reads sum: 4961334 > total (4861702)
Error! Sample 4311900108_S20_R1_ has a bad reads sum: 4507173 > total (3404988)
Error! Sample 0 has a bad reads sum: 2842596 > total (0)
7 errors have been found, these are from samples: ['4311801120_S23_R1_', '4311801213_S7_R1_', '4311801221_S16_R1_', '4311801325_S19_R1_', '4311900015_S21_R1_', '4311900108_S20_R1_', 0]
Wrote read-based table to:	 results/profile_read_counts.csv
Wrote percentages table to:	 results/profile_percentages.csv
Traceback (most recent call last):
  File "/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/scripts/tmp6zz5eq68.quantify_profiles.py", line 512, in <module>
    main()
  File "/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/scripts/tmp6zz5eq68.quantify_profiles.py", line 503, in main
    outfile = snakemake.output["graph"]
  File "/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/scripts/tmp6zz5eq68.quantify_profiles.py", line 306, in draw_stacked_bars
    tooltips="@%s $name: @$name" % sample)
  File "/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/conda/a9f56b99/lib/python3.7/site-packages/bokeh/plotting/figure.py", line 1023, in figure
    return Figure(**kwargs)
  File "/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/conda/a9f56b99/lib/python3.7/site-packages/bokeh/plotting/figure.py", line 191, in __init__
    self.x_range = _get_range(opts.x_range)
  File "/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/conda/a9f56b99/lib/python3.7/site-packages/bokeh/plotting/helpers.py", line 371, in _get_range
    raise ValueError("Unrecognized range input: '%s'" % str(range_input))
ValueError: Unrecognized range input: '['4311801120_S23_R1_' '4311801211_S8_R1_' '4311801213_S7_R1_'
 '4311801221_S16_R1_' '4311801275_S22_R1_' '4311801325_S19_R1_'
 '4311900015_S21_R1_' '4311900108_S20_R1_' 0]'
[Fri May 24 13:59:23 2019]
Error in rule quantify_output:
    jobid: 84
    output: results/profile_read_counts.csv, results/profile_percentages.csv, results/Sample_composition_graph.html
    log: logs/quantify_output.log
    conda-env: /data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/conda/a9f56b99

RuleException:
CalledProcessError in line 660 of /data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/Snakefile:
Command 'source /mnt/miniconda/bin/activate '/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/conda/a9f56b99'; set -euo pipefail;  python /data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/.snakemake/scripts/tmp6zz5eq68.quantify_profiles.py' returned non-zero exit status 1.
  File "/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian/Snakefile", line 660, in __rule_quantify_output
  File "/home/janssetk/.conda/envs/Jovian_master/lib/python3.6/concurrent/futures/thread.py", line 56, in run

```

**Comment by DennisSchmitz** [comment]
Is this on the development branch or on the master branch? Because I know Sam is updating this script on the development branch, so it might be unstable.

You can print the version via `bash jovian -v` to find out

**Comment by thierryjanssens** [comment]
[janssetk@rivm-biohn-l02p Jovian]$ bash jovian -v
Jovian v0.9.3

**Comment by DennisSchmitz** [comment]
I`ll ask Sam to look into it. Could you send him the location of these files via slack? Then he can reproduce it.

**Comment by thierryjanssens** [comment]
/data/BioGrid/ERVINGS/Runs_with_NGS_pool_NExtSEQ_RIVM/Inhouse_NextSeq_20190517_NGS_run20/Jovian

**Comment by samnooij** [comment]
The read sums that turn out higher than the total number of raw reads is not the actual problem. (Perhaps this should be reported as 'warning'.) The script notifies users when sums look wrong, but should still continue to produce output files.

The error seems to be in the `tooltips=...` part. It looks like somehow the variable `sample` is not what it is supposed to be. I'm not sure where that problem would come from... Therefore, it might take a while before it is fixed. If you don't really need these quantification figures, I suggest you ignore it for now (run Jovian with `--keep-going` to finish all other steps).

Also, as Dennis said, I have been working on this script and have also changed the way filenames are parsed. (Technical detail: the script was only tested with sample names ending on '_R1' and '_R2', which were stripped off by removing the final 3 characters. With sample names with a suffix behind R1, this would not work and return unintended sample names as seen in your example. In the new version, the sample name is made be looking for '_R1' and removing that and anything that may be behind it.) Perhaps that would solve this bug as well. I will try a little test with the newer script to see how that works out.

Anyhow, about those summed reads: I checked the all_taxClassified.tsv file and found that for at least one sample, the number of mapped reads is higher than the raw reads. That issue is also being worked on, but it is not as straightforward as I hoped (see #56).

**Comment by DennisSchmitz** [comment]
Sam, I didn't know that was how you implemented it. But @kroonma updated the sample_sheet generator to also work with the ENA filename scheme `[filename]_1.fastq` and with regular '[filename]_R1.fastq' naming scheme. See here `fq_pattern = re.compile("(.*)_R?(1|2).*\.f(ast)?q(\.gz)?")` 

So that's a problem then, if I understand it correctly, that's also the reason that Thierry's error occurs. 
Can this be updated to work with the sample_sheet? Or to use a similar regex as listed above.

**Comment by samnooij** [comment]
Right, using the sample sheet would be nice. I'll see how that works with this script. Good that you mention it! 

While testing the new experimental script I found another bug: counting the number of human reads does not work with empty files. I.e. the script cannot run when there are samples without human reads in them. This should not be too difficult to fix, but it is another pitfall that I did not consider before. Thanks for this test-case, Thierry! (I hope these are all the bugs for now...)
